fendermint/ipc/src/parent.rs

```diff
@@ -0,0 +1,202 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+//! Parent view related functions
+
+use crate::agent::AgentProxy;
+use crate::cache::RangeKeyCache;
+use crate::pof::IPCParentFinality;
+use crate::{BlockHeight, Config, Nonce, ParentViewProvider};
+use anyhow::anyhow;
+use ipc_sdk::cross::CrossMsg;
+use ipc_sdk::ValidatorSet;
+use std::cmp::max;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::{Arc, RwLock};
+use std::time::Duration;
+use tokio::task::JoinHandle;
+
+/// Constantly syncing with parent through polling
+pub struct PollingParentSyncer {
+    config: Config,
+    started: Arc<AtomicBool>,
+    cache: LockedCache,
+    ipc_agent_proxy: Arc<AgentProxy>,
+    handle: Option<JoinHandle<anyhow::Result<()>>>,
+}
+
+impl PollingParentSyncer {
+    pub fn new(config: Config) -> Self {
+        Self {
+            config,
+            started: Arc::new(AtomicBool::new(false)),
+            cache: Arc::new(RwLock::new(ParentSyncerCache {
+                block_hash: RangeKeyCache::new(),
+                top_down_message: RangeKeyCache::new(),
+                membership: None,
+            })),
+            ipc_agent_proxy: Arc::new(AgentProxy {}),
+            handle: None,
+        }
+    }
+}
+
+impl ParentViewProvider for PollingParentSyncer {
+    fn latest_height(&self) -> Option<BlockHeight> {
+        self.read_cache(|cache| cache.block_hash.upper_bound())
+    }
+
+    fn block_hash(&self, height: BlockHeight) -> Option<Vec<u8>> {
+        self.read_cache(|cache| cache.block_hash.get_value(height).map(|v| v.to_vec()))
+    }
+
+    fn top_down_msgs(&self, _height: BlockHeight, nonce: Nonce) -> Vec<CrossMsg> {
+        self.read_cache(|cache| {
+            let v = cache.top_down_message.values_within_range(nonce, None);
+            // FIXME: avoid clone here, return references, let caller clone on demand
+            v.into_iter().cloned().collect()
+        })
+    }
+
+    fn membership(&self) -> Option<ValidatorSet> {
+        self.read_cache(|cache| cache.membership.clone())
+    }
+
+    fn on_finality_committed(&self, finality: &IPCParentFinality) {
```

Comment:
So here we remove all entries from the cache up to the `finality` height, which is calculated elsewhere as `latest - delay`. Yet the parent fetcher will fill data in each loop between `latest - lower_bound` and `latest`. Doesn't that mean that there is, or should be no difference between `delay` and `lower_bound`, otherwise one one deletes, the other puts back. Or actually it can't put it back because of the cache cutoff, but it still keeps fetching it.

Reply:
This dissonance would disappear if you only kept one value, the finality-delay, and in the fetcher you:
1. probed the last height you fetched before
2. if it's the same as before, fetch the next region up to the latest height
3. if it's not the same, find the greatest common ancestor (walk backwards either linearly, or skipping), and then re-fetch up to the latest height, and replace

That way you only fetch what you need, and never anything beyond the finality height (if it was indeed final). Under these conditions the cache cutoffs wouldn't be required, because if the configuration is right, you should never fetch lower than lowest data.

Reply:
Similarly upon startup, you can initialize the component with the last known finalized block, then check if it's still the same on the parent: if it doesn't exist, the parent is syncing; if it's different, it's a fatal error because the subnet state diverged from the parent. Then, go fetch until the latest height is reached, but be careful that we only start doing this once our own node is caught up with the network, or you might be fetching the entire parent chain (if our own node is going from genesis). 

Once you have fetched all the region between the last known finalized block and the latest, then you can use the finality-delay to choose the next proposal.